presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryLeakDetector.java

```diff
@@ -0,0 +1,160 @@
+import static java.util.Collections.newSetFromMap;
+import static java.util.Objects.requireNonNull;
+import static org.joda.time.DateTime.now;
+import static org.joda.time.Seconds.secondsBetween;
+
+@ThreadSafe
+public class ClusterMemoryLeakDetector
+{
+    private static final Logger log = Logger.get(ClusterMemoryLeakDetector.class);
+
+    // It may take some time to remove a query's memory reservations from the worker nodes, that's why
+    // we check to see whether some time has passed after the query finishes to claim that it is leaked.
+    private static final int DEFAULT_LEAK_CLAIM_DELTA_SEC = 60;
+
+    // By default, leakedQueries will be a Set (created from a WeakHashMap -- just to be safe in case map gets large) of leaked query IDs.
+    @GuardedBy("this")
+    private final LeakedQueries leakedQueries;
+
+    // We don't rely on the size of the weak leakedQueries map for the number of leaked queries as we want an exact counter.
+    private final AtomicInteger numberOfLeakedQueries = new AtomicInteger();
+
+    static ClusterMemoryLeakDetector createMemoryLeakDetector()
+    {
+        return new ClusterMemoryLeakDetector(new LeakedQueries(true));
+    }
+
+    @VisibleForTesting
+    static ClusterMemoryLeakDetector createTestMemoryLeakDetector()
+    {
+        // For testing we want strong references, otherwise the tests will be non-deterministic due to GC effects.
+        return new ClusterMemoryLeakDetector(new LeakedQueries(false));
+    }
+
+    private ClusterMemoryLeakDetector(LeakedQueries leakedQueries)
+    {
+        this.leakedQueries = requireNonNull(leakedQueries, "leakedQueries is null");
+    }
+
+    /**
+     * @param queryExecutionsSupplier All queries that the coordinator knows about.
+     * @param queryMemoryReservations The memory reservations of queries in the GENERAL cluster memory pool.
+     */
+    void checkForMemoryLeaks(Supplier<List<QueryInfo>> queryInfoSupplier, Map<QueryId, Long> queryMemoryReservations)
+    {
+        requireNonNull(queryInfoSupplier);
+        requireNonNull(queryMemoryReservations);
+
+        Map<QueryId, Long> leakedQueryReservations = queryMemoryReservations.entrySet()
+                .stream()
+                .filter(Objects::nonNull)
+                .filter(entry -> entry.getValue() > 0)
+                .filter(entry -> isLeaked(queryInfoSupplier, entry.getKey()))
+                .collect(toImmutableMap(Entry::getKey, Entry::getValue));
+
+        if (!leakedQueryReservations.isEmpty()) {
+            log.warn("Memory leak detected. The following queries are already finished, " +
+                    "but they have memory reservations on some worker node(s): %s", leakedQueryReservations);
+        }
+
+        // keep numberOfLeakedQueries and leakedQueries in sync
+        synchronized (this) {
+            numberOfLeakedQueries.set(leakedQueryReservations.size());
+            leakedQueries.update(leakedQueryReservations.keySet());
```

Comment:
Have you considered deduping in `leakedQueries` as best as possible, and then incrementing `numberOfLeakedQueries` only by the deduplicated number?

Reply:
`leakedQueries` is a set, so there won't be duplicates.

Reply:
I meant duplicate reporting on subsequent runs